### ISSUE TITLE
Mangle property names

### ISSUE THREAD

**Issue by Rich-Harris** [Rich-Harris/butternut#10]
This can have a substantial impact on generated code size. Would like it to be automated though — no annotations or white/blacklists, instead Butternut should ideally be smart enough to figure out whether a given property is accessible from the outside world (either because it's part of an exported object, returned from an exported function, or is part of the environment). Definitely not trivial, but would be neat

**Comment by pygy** [comment]
Another corner case is the argument/context of DOM event handlers (`e.currentTarget`, etc...).

**Comment by Rich-Harris** [comment]
@pygy Could you elaborate?

**Comment by pygy** [comment]
@Rich-Harris More generally, objects that are injected from foreign code as function arguments must not have their properties mangled.

Maybe that was implicitly covered by "part of an exported object"?

**Comment by Rich-Harris** [comment]
Yes — anything that isn't defined inside the bundle would be considered strictly off-limits. The theory is that it would be able to trace the object that the property belongs to back to a declaration or literal inside the source code:

```js
// this...
function logCurrentTarget (e) {
  console.log(e.currentTarget);
}

logCurrentTarget({currentTarget: 42});

// ...would become this...
function logCurrentTarget(e){console.log(e.a)}logCurrentTarget({a: 42})

// ...but this...
function logCurrentTarget (e) {
  console.log(e.currentTarget);
}

logCurrentTarget({currentTarget: 42});
window.addEventListener('click', logCurrentTarget);

// would become this, because we can no longer determine where the arguments
// to `logCurrentTarget` originated from:
function logCurrentTarget(e){console.log(e.currentTarget)}logCurrentTarget({currentTarget:
42})window.addEventListener('click',logCurrentTarget)
```

It's a *little* easier said than done, but in theory I think it's possible.
